File: examples/upload_client.py
#!/usr/bin/env python3
"""
文件上传客户端示例

使用方法：
python upload_client_example.py <file_path>

示例：
python upload_client_example.py test.txt
"""

import hmac
import hashlib
import os
import time
import sys
import requests

# API配置
API_KEY = os.getenv("AUTH_API_KEY", "your-secret-api-key-here")  # 需要与服务器端的ApiKey一致
SERVER_URL = os.getenv("SERVER_URL", "http://localhost:9012/upload")  # 或 https://localhost:9013/upload


def calculate_signature(api_key, timestamp):
    """计算HMAC-SHA256签名"""
    h = hmac.new(api_key.encode(), timestamp.encode(), hashlib.sha256)
    return h.hexdigest()


def upload_file(file_path, upload_dir=None):
    """上传文件到服务器"""
    # 生成时间戳
    timestamp = str(int(time.time()))
    
    # 计算签名
    signature = calculate_signature(API_KEY, timestamp)
    
    # 设置请求头
    headers = {
        'X-Timestamp': timestamp,
        'X-Signature': signature,
    }
    
    # 准备文件
    try:
        with open(file_path, 'rb') as f:
            files = {'file': f}
            
            # 构建URL（可选：指定上传目录）
            url = SERVER_URL
            if upload_dir:
                url = f"{SERVER_URL}?dir={upload_dir}"
            
            # 发送请求
            response = requests.post(url, files=files, headers=headers)
            
            # 处理响应
            if response.status_code == 200:
                result = response.json()
                print(f"✅ 上传成功!")
                print(f"   文件名: {result.get('filename')}")
                print(f"   大小: {result.get('size')} bytes")
                print(f"   保存路径: {result.get('path')}")
            else:
                print(f"❌ 上传失败!")
                print(f"   状态码: {response.status_code}")
                print(f"   错误信息: {response.text}")
                
    except FileNotFoundError:
        print(f"❌ 文件不存在: {file_path}")
    except Exception as e:
        print(f"❌ 发生错误: {e}")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("使用方法: python upload_client_example.py <file_path> [upload_dir]")
        print("示例: python upload_client_example.py test.txt")
        print("示例: python upload_client_example.py test.txt ./custom_uploads")
        sys.exit(1)
    
    file_path = sys.argv[1]
    upload_dir = sys.argv[2] if len(sys.argv) > 2 else None
    
    upload_file(file_path, upload_dir)

File: examples/upload_client.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"time"
)

const (
	// API配置 - 需要与服务器端的ApiKey一致
	ApiKey    = "your-secret-api-key-here"
	ServerURL = "http://localhost:9012/upload" // 或 https://localhost:9013/upload
)

// UploadResponse 服务器响应结构
type UploadResponse struct {
	Success  bool   `json:"success"`
	Filename string `json:"filename"`
	Size     int64  `json:"size"`
	Path     string `json:"path"`
	Error    string `json:"error,omitempty"`
}

// calculateHMAC 计算HMAC-SHA256签名
func calculateHMAC(key, message string) string {
	h := hmac.New(sha256.New, []byte(key))
	h.Write([]byte(message))
	return hex.EncodeToString(h.Sum(nil))
}

// uploadFile 上传文件到服务器
func uploadFile(filePath string, uploadDir string) error {
	// 打开文件
	file, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("打开文件失败: %w", err)
	}
	defer file.Close()

	// 创建multipart表单
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	// 添加文件字段
	part, err := writer.CreateFormFile("file", filepath.Base(filePath))
	if err != nil {
		return fmt.Errorf("创建表单文件字段失败: %w", err)
	}

	// 复制文件内容
	if _, err := io.Copy(part, file); err != nil {
		return fmt.Errorf("复制文件内容失败: %w", err)
	}

	// 关闭writer
	if err := writer.Close(); err != nil {
		return fmt.Errorf("关闭writer失败: %w", err)
	}

	// 生成时间戳
	timestamp := strconv.FormatInt(time.Now().Unix(), 10)

	// 计算签名
	signature := calculateHMAC(ApiKey, timestamp)

	// 构建URL
	url := ServerURL
	if uploadDir != "" {
		url = fmt.Sprintf("%s?dir=%s", ServerURL, uploadDir)
	}

	// 创建请求
	req, err := http.NewRequest("POST", url, body)
	if err != nil {
		return fmt.Errorf("创建请求失败: %w", err)
	}

	// 设置请求头
	req.Header.Set("Content-Type", writer.FormDataContentType())
	req.Header.Set("X-Timestamp", timestamp)
	req.Header.Set("X-Signature", signature)

	// 发送请求
	client := &http.Client{
		Timeout: 30 * time.Second,
	}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("发送请求失败: %w", err)
	}
	defer resp.Body.Close()

	// 读取响应
	var uploadResp UploadResponse
	if err := json.NewDecoder(resp.Body).Decode(&uploadResp); err != nil {
		return fmt.Errorf("解析响应失败: %w", err)
	}

	// 处理响应
	if resp.StatusCode == http.StatusOK && uploadResp.Success {
		fmt.Printf("✅ 上传成功!\n")
		fmt.Printf("   文件名: %s\n", uploadResp.Filename)
		fmt.Printf("   大小: %d bytes\n", uploadResp.Size)
		fmt.Printf("   相对路径: %s\n", uploadResp.Path)
		fmt.Printf("   (相对于根上传目录)\n")
		return nil
	} else {
		return fmt.Errorf("上传失败 (状态码: %d): %s", resp.StatusCode, uploadResp.Error)
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("使用方法: go run upload_client_example.go <file_path> [upload_dir]")
		fmt.Println("示例: go run upload_client_example.go test.txt")
		fmt.Println("示例: go run upload_client_example.go test.txt ./custom_uploads")
		os.Exit(1)
	}

	filePath := os.Args[1]
	uploadDir := ""
	if len(os.Args) > 2 {
		uploadDir = os.Args[2]
	}

	if err := uploadFile(filePath, uploadDir); err != nil {
		fmt.Printf("❌ 错误: %v\n", err)
		os.Exit(1)
	}
}
